### PR TITLE
Suggested keywords

### PR DESCRIPTION
**Pull request by Lenninlasd**
## What does this PR do?
- [x] Add a each search in the input as a keyword.
- [x] Delete keywords by clicking on a close button in the search box.
- [x] Delete a keyword when press the backspace key button and the input is clean.
- [x] Refresh list when zooming or filtering by other means.
- [x] Integrate keyword with API
- [x] When clicking on the keywords they should get added to the search terms.
- [x] Test! test! test!
### Screenshots:

<div align="center">
  <span>

  </span>
  <span>

  </span>
</div>

### Related Issue

https://github.com/terranodo/angular-search/issues/42

### CODE REVIEW

File: app/components/keyword/keyword.tpl.html [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<div class="md-chips">
+    <span>{{tag}}</span>
+    <button ng-click="removefn(tag)" ng-if="remove">&times</button>
```
Comment (milafrerichs) [inline app/components/keyword/keyword.tpl.html:3]: if you want to use `&` please use `&amp;`

File: app/components/keyword/keywordDirective.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+(function() {
+    angular
+    .module('search_keyword_component', [])
+    .directive('keyword', function keyword() {
+        return {
+            scope: {
+                tag: '@',
+                remove: '=',
+                removefn: '='
```
Comment (milafrerichs) [inline app/components/keyword/keywordDirective.js:9]: please use `&` when you want to call a function. 
do not pass the function in the scope.

File: app/components/keyword/keywordDirective.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+(function() {
+    angular
+    .module('search_keyword_component', [])
+    .directive('keyword', function keyword() {
+        return {
+            scope: {
+                tag: '@',
+                remove: '=',
```
Comment (milafrerichs) [inline app/components/keyword/keywordDirective.js:8]: why is there a boolean for if the tag is removed or not? could you just remove it from the scope?

File: app/components/toolbarSearch/toolbarSearchDirective.js [hunk 4]
```diff
@@ -36,34 +54,77 @@
...
+                    vm.textSearchInput.previousLength = vm.textSearchInput.value.length;
+                }
+
+                function addSuggestedKeywordToSearchInput(keyword) {
+                    vm.textSearchInput.value = keyword;
+                    doSearch();
+                }
+
+                function listenSuggestWords() {
+                    vm.$on('setSuggestWords', function(event, dataRawKeywords) {
```
Comment (milafrerichs) [inline app/components/toolbarSearch/toolbarSearchDirective.js:122]: we can remove this watcher if you take a look at how I structred the `filter`

File: app/components/toolbarSearch/toolbarSearchDirective.js [hunk 5]
```diff
@@ -36,34 +54,77 @@
...
+                }
+
+                function keyWordStringToArray(keyWordString) {
+                    keyWordString = keyWordString || '';
+                    return keyWordString.split('"').filter(function(val){
+                        return val !== '' && val !== ' ';
+                    });
+                }
+
+                function removeKeyWord(keyword) {
```
Comment (milafrerichs) [inline app/components/toolbarSearch/toolbarSearchDirective.js:97]: I don't quite understand this method yet. Can you explain a little more?

File: app/components/toolbarSearch/toolbarSearchDirective.js [hunk 6]
```diff
@@ -19,12 +20,29 @@
...
                 var vm = scope;
-                /**
-                 *
-                 */
+
+
                 vm.filter = searchFilter;
+                vm.filterArray = [];
+                vm.suggestedKeywords = [];
+                vm.textSearchInput = {value: '', previousLength: 0 };
```
Comment (milafrerichs) [inline app/components/toolbarSearch/toolbarSearchDirective.js:29]: can you explain the usage if this variable as well? Not sure where previousLength comes into play.

File: app/components/keyword/keyword.tpl.html (hunk 1 above)
Comment (Lenninlasd) [reply app/components/keyword/keyword.tpl.html:3]: I corrected by html entity `&times;`

File: app/components/keyword/keywordDirective.js (hunk 3 above)
Comment (Lenninlasd) [reply app/components/keyword/keywordDirective.js:8]: To hide or show the remove option from outside

File: app/components/toolbarSearch/toolbarSearchDirective.js (hunk 6 above)
Comment (Lenninlasd) [reply app/components/toolbarSearch/toolbarSearchDirective.js:29]: In the case we need erase the last character of the input, and not delete the last keyword of the search

File: app/components/keyword/keywordDirective.js (hunk 2 above)
Comment (Lenninlasd) [reply app/components/keyword/keywordDirective.js:9]: Fixed

File: app/components/toolbarSearch/toolbarSearchDirective.js (hunk 5 above)
Comment (Lenninlasd) [reply app/components/toolbarSearch/toolbarSearchDirective.js:97]: If the list of keywords in the search is ['test1', 'test2', 'test3'], and we want delete 'test2', we pass this value and the and search in HeatMapSourceGenerator `'"test1" "test3"'`

File: app/components/toolbarSearch/toolbarSearchDirective.js (hunk 4 above)
Comment (Lenninlasd) [reply app/components/toolbarSearch/toolbarSearchDirective.js:122]: Got it, I think we need a promise or a callback that return the requested information

**Pull request merged**